Add  lps22df sensor driver (required for sensortile.box)

avisconti opened the pull request:
Add support to LPS22DF sensors which is going to be used later to extend sensortile.box pro board.
Based on STdC API v2.3

Please note that lis2dux12/lis2duxs12 sensor drivers have been removed in order to possibly speed up PR review.

EDIT:
Added also sensor drivers build test part for I3C bus.

avisconti committed:
drivers/sensor: add support to LPS22DF pressure sensor

The LPS22DF is an ultracompact, piezoresistive, absolute pressure sensor
that functions as a digital output barometer. The LPS22DF provides lower
power consumption, achieving lower pressure noise than its predecessor.
This driver is based on stmemsc HAL i/f v2.3

https://www.st.com/en/datasheet/lps22df.pdf

Signed-off-by: Armando Visconti <armando.visconti@st.com>

erwango commented on dts/bindings/sensor/st,lps22df-common.yaml:
```diff
@@ -0,0 +1,65 @@
+        Specify the output data rate expressed in samples per second (Hz).
+        The default is the power-on reset value.
+    enum:
+      - 0  # Power-Down
+      - 1  # 1Hz
+      - 2  # 4Hz
+      - 3  # 10Hz
+      - 4  # 25Hz
+      - 5  # 50Hz
+      - 6  # 75Hz
+      - 7  # 100Hz
+      - 8  # 200Hz
```
Description need to be updated here as well

avisconti replied:
NOt sure why i'm so stupid

avisconti commented:
rebased and repushed to resolve conflicts

avisconti commented:
rebased and repushed again to solve conflicts.
@erwango @MaureenHelm can you possibly review it once again? Sorry, but the test part gets in conflict often as new sensor are added in main. Thanks

MaureenHelm commented:
`LIS2DUXS12` and `LIS2DUX12` seem to have a lot in common. The part numbers are so similar that at first I didn't see the sneaky S in the middle that distinguishes them from each other.

Is it possible to support both parts with one driver to avoid duplicating code?

avisconti commented:
> `LIS2DUXS12` and `LIS2DUX12` seem to have a lot in common. The part numbers are so similar that at first I didn't see the sneaky S in the middle that distinguishes them from each other.
> 
> Is it possible to support both parts with one driver to avoid duplicating code?

I'll try to see if it is possible to treat them as variants of the same device, hence sharing a single driver

avisconti commented:
> `LIS2DUXS12` and `LIS2DUX12` seem to have a lot in common. The part numbers are so similar that at first I didn't see the sneaky S in the middle that distinguishes them from each other.
> 
> Is it possible to support both parts with one driver to avoid duplicating code?

I have currently removed lis2dux12 sensor driver. I'll think about how to handle it in same driver later on.

avisconti committed:
boards: arm: sensortile_box_pro: extend with lps22df

Extend sensortile_box_pro with lps22df.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
sample: board: sensortile_box_pro: extend with lps22df

Extend sensor sample reading also lps22df data.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

jremmert-phytec-iot commented on drivers/sensor/lps22df/lps22df_trigger.c:
```diff
@@ -0,0 +1,256 @@
+		return;
+	}
+
+	if (status.drdy_pres != 0) {
```
I can see some problems here when I tested an adapted version of this PR for an lps28dfw pressure sensor.

`s/if (status.drdy_pres != 0)/if (!status.drdy)`
`status.drdy_pres` should be `1` (status reg; P_DA bit) if pressure data is available.

The pin interrupt gets deactivated in `lps22df_gpio_callback` before this function is called. In case we have a spurious interrupt, we should probably configure the interrupt pin so that we will get future interrupts.

avisconti replied:
> I can see some problems here when I tested an adapted version of this PR for an lps28dfw pressure sensor.
> 
> `s/if (status.drdy_pres != 0)/if (!status.drdy)` `status.drdy_pres` should be `1` (status reg; P_DA bit) if pressure data is available.
> 
> The pin interrupt gets deactivated in `lps22df_gpio_callback` before this function is called. In case we have a spurious interrupt, we should probably configure the interrupt pin so that we will get future interrupts.

Yes, sure. Good catch!

I may add an exit: strategy at the bottom:

```
        if (status.drdy_pres != 0) {
                goto exit; /* spurious interrupt */
        }

exit:
        ret = gpio_pin_interrupt_configure_dt(&cfg->gpio_int,
                                              GPIO_INT_EDGE_TO_ACTIVE);
        if (ret < 0) {
                LOG_ERR("%s: Not able to configure pin_int", dev->name);
        }

avisconti replied:
done

jremmert-phytec-iot replied:
The exit strategy looks good, but the drdy_pres bit is still inverted? If drdy_pres is 1, data is available and therefore we do not have a spurious interrupt.

avisconti replied:
> The exit strategy looks good, but the drdy_pres bit is still inverted? If drdy_pres is 1, data is available and therefore we do not have a spurious interrupt.

Yes sure... not sure how I can be so stupid. Thanks very much for being so patient...

avisconti replied:
hope it might be a good starting point to extend the support to LPS28DFW:
https://github.com/zephyrproject-rtos/zephyr/issues/64766

fabiobaltieri commented on drivers/sensor/lps22df/lps22df.c:
```diff
@@ -0,0 +1,356 @@
+static inline int lps22df_set_odr_raw(const struct device *dev, uint8_t odr)
+{
+	const struct lps22df_config * const cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
```
Do you need the casting?

```c
struct lps22df_config {
	stmdev_ctx_t ctx;
...
```

avisconti replied:
Yes, it's just to avoid this warning:

```
warning: initialization discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
   76 |         stmdev_ctx_t *ctx = &cfg->ctx;
```

avisconti replied:
@fabiobaltieri it is something that was highlighted some times ago. In order to fix it I guess I would need to change all the stmemscc module APIs adding const to ctx. Like this:

```
int32_t lps22df_bus_mode_set(const stmdev_ctx_t *ctx, lps22df_bus_mode_t *val)```
```

STdC are something that I can control, but it is a huge change, even if good change (ctx is really not modifiable). I will think about it.

fabiobaltieri replied:
got it, yeah not worth it

fabiobaltieri commented on drivers/sensor/lps22df/lps22df.c:
```diff
@@ -0,0 +1,356 @@
+{
+	const struct lps22df_config * const cfg = dev->config;
+	struct lps22df_data *data = dev->data;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
```
same

avisconti replied:
Yeah, I'll think about changing the STdC APIs for all modules sometimes in the future. After that I should be able to get rid of this casting.

fabiobaltieri commented on drivers/sensor/lps22df/lps22df.c:
```diff
@@ -0,0 +1,356 @@
+			LOG_DBG("sw reset timed out");
+			return -ETIMEDOUT;
+		}
+		k_usleep(50);
```
add a define for 50

avisconti replied:
ack

fabiobaltieri commented on drivers/sensor/lps22df/lps22df.c:
```diff
@@ -0,0 +1,356 @@
+		}
+	}
+#else
+	ARG_UNUSED(data);
```
`ARG_UNUSED` is typically for arguments as the name implies, use `__maybe_unused` in the local variable declaration instead

avisconti replied:
ok, thanks. Looks a better choice infact.

fabiobaltieri commented on drivers/sensor/lps22df/lps22df.c:
```diff
@@ -0,0 +1,356 @@
+	return 0;
+}
+
+#if DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 0
+#warning "LPS22DF driver enabled without any devices"
+#endif
```
I don't think this can ever happen since you have

```
depends on DT_HAS_ST_LPS22DF_ENABLED
```

so you may as well drop it

avisconti replied:
that's correct, thanks

fabiobaltieri commented on drivers/sensor/lps22df/lps22df.c:
```diff
@@ -0,0 +1,356 @@
+	.lpf = DT_INST_PROP(inst, lpf),					\
+	.avg = DT_INST_PROP(inst, avg),					\
+	.drdy_pulsed = DT_INST_PROP(inst, drdy_pulsed),			\
+	COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, drdy_gpios),		\
+			(LPS22DF_CFG_IRQ(inst)), ())
```
nit: you'd  normally align on the parenthesis above

```c
	COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, drdy_gpios),		\
		    (LPS22DF_CFG_IRQ(inst)), ())
```

fabiobaltieri commented on drivers/sensor/lps22df/lps22df.c:
```diff
@@ -0,0 +1,356 @@
+	COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, drdy_gpios),		\
+			(LPS22DF_CFG_IRQ(inst)), ())
+
+#define LPS22DF_SPI_OPERATION (SPI_WORD_SET(8) |			\
+				SPI_OP_MODE_MASTER |			\
+				SPI_MODE_CPOL |				\
+				SPI_MODE_CPHA)				\
```
same, plus this could fit in half the lines if you want to keep it a bir mote compact

```c
#define LPS22DF_SPI_OPERATION (SPI_WORD_SET(8) | SPI_OP_MODE_MASTER |   \
                               SPI_MODE_CPOL | SPI_MODE_CPHA)           \
```

fabiobaltieri commented on drivers/sensor/lps22df/lps22df_trigger.c:
```diff
@@ -0,0 +1,257 @@
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	lps22df_data_t raw_data;
+
+	if (trig->chan == SENSOR_CHAN_ALL) {
```
swap the logic here,

```c
if (trig->chan != SENSOR_CHAN_ALL) {
        return -ENOTSUP;
}

<normal execution inline>
```

easies rto track the intended code path

avisconti replied:
seems proper

fabiobaltieri commented on drivers/sensor/lps22df/lps22df_trigger.c:
```diff
@@ -0,0 +1,257 @@
+	int ret;
+
+	/* setup data ready gpio interrupt */
+	if (!device_is_ready(cfg->gpio_int.port)
```
use `gpio_is_ready_dt`

fabiobaltieri commented on drivers/sensor/lps22df/lps22df_trigger.c:
```diff
@@ -0,0 +1,257 @@
+	lps22df->work.handler = lps22df_work_cb;
+#endif /* CONFIG_LPS22DF_TRIGGER_OWN_THREAD */
+
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i3c)
+	if (cfg->i3c.bus == NULL)
+#endif
```
these are quite a bit in the way, how about having a `#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i3c)`  on the top and maybe setting a define as either a condition or true?

```c
#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i3c)
        #define ON_I3C_BUS(cfg) (cfg->i3c.bus != NULL)
#else
        #define ON_I3C_BUS false
#endif
```

and then these would be plain `if (!ON_I3C_BUS(cfg)) {`

avisconti replied:
Yes, thanks for suggestion: it will improve a lot the code readibility.
Moreover I saw some other issues on I3C part, so that I will review it all.
Thanks.

fabiobaltieri commented on tests/drivers/build_all/sensor/spi.dtsi:
probably worth starting an `i3c.dtsi` somewhere here if there's not already, so you'd get the conditional checked too

avisconti replied:
Not there already. But good point!

avisconti replied:
ok, done!

fabiobaltieri commented on samples/boards/sensortile_box_pro/sensors-on-board/src/main.c:
```diff
@@ -266,6 +316,14 @@ int main(void)
+		       sensor_value_to_double(&lps22df_temp));
+
+		/* pressure */
+		printf("LPS22DF: Pressure:%.3f kpa\n",
```
nit: `Pressure: %.3f kpa\n` (add a space, the other have it)

avisconti committed:
drivers: i3c: add dummy driver for vnd,i3c

It is just used to be able to DEVICE_DT_GET() bus devices from
tests/drivers/build_all in an upcoming commit.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
tests: drivers: build_all: sensor: add I3C test framework

Add framework to test sensor drivers I3C conditional checkings.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

fabiobaltieri commented on drivers/i3c/i3c_test.c:
```diff
@@ -0,0 +1,48 @@
+
+static int vnd_i3c_configure(const struct device *dev,
+			     enum i3c_config_type type, void *config)
+
```
nit: drop the blank line

fabiobaltieri commented on dts/bindings/sensor/st,lps22df-common.yaml:
```diff
@@ -0,0 +1,88 @@
+        Specify the output data rate expressed in samples per second (Hz).
+        The default is the power-on reset value.
+
+        0 = Power-Down
+        1 = 1Hz
+        2 = 4Hz
+        3 = 10Hz
+        4 = 25Hz
+        5 = 50Hz
+        6 = 75Hz
+        7 = 100Hz
+        8 = 200Hz
+    enum:
+      - 0
+      - 1
+      - 2
+      - 3
+      - 4
+      - 5
+      - 6
+      - 7
+      - 8
```
not for this PR since this is coherent with other ST drivers, but if you or anyone from ST feels like doing a swipe of the other ST bindings in the future: I think it'd be a nice usability improvements to have a matching set of `#define`s for these in `include/zephyr/dt-bindings/sensor`, one file per class of device I guess.

Then the dt instance of these would look something like `odr = <LPS22DF_ODR_100HZ>;` rather than `odr = <7>;`, one would not have to cross reference the binding documentation, you could change the description of this to just say "any of LPS22DF_ODR_..." etc...

avisconti replied:
> not for this PR since this is coherent with other ST drivers, but if you or anyone from ST feels like doing a swipe of the other ST bindings in the future: I think it'd be a nice usability improvements to have a matching set of `#define`s for these in `include/zephyr/dt-bindings/sensor`, one file per class of device I guess.
> 

Sure, I can think about it. Probably I would be the person that would do it, and I can do a pass for all ST devices in a single PR.

> Then the dt instance of these would look something like `odr = <LPS22DF_ODR_100HZ>;` rather than `odr = <7>;`, one would not have to cross reference the binding documentation, you could change the description of this to just say "any of LPS22DF_ODR_..." etc...

Exactly. At that point there would be no need for explain each case one by one, as the name itself would describe it. And to be hones it is very bad to read now.

fabiobaltieri replied:
Yeah, cross referencing is painful

fabiobaltieri reviewed (commented):
good stuff, just one nitpick and one comment for a potential followup

avisconti commented:
@erwango

jonas-rem reviewed (approved):
I tested a slightly modified version of this PR with two ST LPS28DFW sensors, connected to a reel board via I2C and two GPIOs for Interrupt.

Will open a PR of a driver version for the LPS28DFW once this PR is merged.

zafersn commented:
Hi @avisconti @fabiobaltieri. I have just luckily come across this issue. I too worked on this sensor driver and wanted to publish it as a zephyros driver but you were quicker. anyway, I have a couple of questions for you if you don't mind to understand your approach to how a zephyros driver should be comprehensive. 


**question 1:** 

        if (ON_I3C_BUS(cfg)) {
                lps22df_bus_mode_t bus_mode;


                /* Select bus interface */
                bus_mode.filter = LPS22DF_AUTO;
                bus_mode.interface = LPS22DF_SEL_BY_HW;
                lps22df_bus_mode_set(ctx, &bus_mode);


        }

here is your bus mode setting, may I ask why you restrict bus_mode_settings with only the i3c bus interface? Because this bus mode setting is not only for i3c but also is valid for  I²C with antispike filter as well, if you are running this code for i3c you should run it for the i2c bus too. 

look at the lps22df.pdf  page 28, paragraph: 7.4.2:

> Overview of antispike filter management
> The device acts as a standard I²C target as long as it has an I²C static address. The device is capable of
> detecting and disabling the I²C antispike filter after detecting the broadcast address (7'h7E/W). In order to
> guarantee proper behavior of the device, the I3C master must emit the first START, 7'h7E/W at open-drain speed
> using I²C fast mode plus reference timing.
> After detecting the broadcast address, the device can receive the I3C dynamic address following the I3C pushpull timing. If the device is not assigned a dynamic address, then the device continues to operate as an I²C device
> with no antispike filter. **For the case in which the host decides to keep the device as I²C with antispike filter**, there
> is a configuration required to keep the antispike filter active. This configuration is done by writing the ASF_ON bit
> to 1 in the I3C_IF_CTRL_ADD (19h) register. This configuration forces the antispike filter to always be turned on
> instead of being managed by the communication on the bus.
> 

```
#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i3c)
        #define ON_I3C_BUS(cfg) (cfg->i3c.bus != NULL)
#else
        #define ON_I3C_BUS(cfg) (false)
#endif
```

but look at your configuration it is only valid if the bus interface is i3c. 

so would it not be better to remove if(ON_I3C_BUS(cfg)) statement?


**question 2:** 

```

/* set sensor default odr */
	LOG_DBG("%s: odr: %d", dev->name, cfg->odr);
	ret = lps22df_set_odr_raw(dev, cfg->odr);
	if (ret < 0) {
		LOG_ERR("%s: Failed to set odr %d", dev->name, cfg->odr);
		return ret;
	}
```

may I ask why you call this function set_odr or set only odr setting here? Because the function lps22df_mode_set you used in the lps22df_set_odr_raw requires also cfg->avg and cfg->lpf configurations which are really important together in terms of power consumption,(see page 33, paragraph 9.6 ) so would be better to give these parameters together into the function you defined (lps22df_set_odr_raw) and rename it. I see you call this function in the lps22df_attr_set to change sampling frequency but the avg(resolution ) config is also very important. I would've added the SENSOR_ATTR_OVERSAMPLING factor with the SENSOR_ATTR_SAMPLING_FREQUENCY in the lps22df_attr_set function to make this sensor fully functional.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/13236364/07da9c9f-12f1-4215-8c52-ac29951f685e)


it seems you took the  lps22hh driver as a sample, that's why you have limited configuration here but lps22hh sensor supports only the odr setting as a sensor attribute but this lps22df sensor has some more attributes 


**question 3**

  should not it be better to have this tries value as a config definition? As far as I understand here this one is the timeout counter and waits until the reset is completed. if the reset is not completed, it is an error. so you defined the tries variable and its value as 10 in the init function in the source file

```

do {
		if (!--tries) {
			LOG_DBG("sw reset timed out");
			return -ETIMEDOUT;
		}
		k_usleep(LPS22DF_SWRESET_WAIT_TIME);

		if (lps22df_status_get(ctx, &status) < 0) {
			return -EIO;
		}
	} while (status.sw_reset);
```

**question 4:** 

we have 2 additional interrupt configs for the lps22df sensor.
would it not be better to show these configurations here to a user of this driver?

```
uint8_t int_latched  : 1; /* int events are: int on threshold, FIFO */
uint8_t active_low   : 1; /* 1 = active low / 0 = active high */
```

```
/* enable drdy in pulsed/latched mode */
	LOG_DBG("drdy_pulsed is %d", (int)cfg->drdy_pulsed);
	mode.drdy_latched = ~cfg->drdy_pulsed;
	if (lps22df_interrupt_mode_set(ctx, &mode) < 0) {
		return -EIO;
	}
```

thanks for reading my questions patiently. and I'll appreciate it if you can reply to me

fabiobaltieri commented:
> thanks for reading my questions patiently. and I'll appreciate it if you can reply to me

Hey could open an issue with it? https://github.com/zephyrproject-rtos/zephyr/issues/new/choose that's a better place for keep the discussion going

zafersn commented:
> > thanks for reading my questions patiently. and I'll appreciate it if you can reply to me
> 
> Hey could open an issue with it? https://github.com/zephyrproject-rtos/zephyr/issues/new/choose that's a better place for keep the discussion going and loop in the ST folks

sorry I'm quite new here, what should I choose as a template? Enhancement?

fabiobaltieri commented:
> sorry I'm quite new here, what should I choose as a template? Enhancement?

Sounds alright, we can always re-tag it.
